File: bottomsheets-sheets/src/main/java/com/arthurivanets/bottomsheets/sheets/ActionPickerBottomSheet.java
/*
 * Copyright 2017 Arthur Ivanets, arthur.ivanets.l@gmail.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.arthurivanets.bottomsheets.sheets;

import android.annotation.SuppressLint;
import android.app.Activity;

import com.arthurivanets.adapster.util.Preconditions;
import com.arthurivanets.bottomsheets.sheets.adapters.actionpicker.ActionItem;
import com.arthurivanets.bottomsheets.sheets.adapters.actionpicker.ActionItemViewHolder;
import com.arthurivanets.bottomsheets.sheets.adapters.actionpicker.ActionPickerItemResources;
import com.arthurivanets.bottomsheets.sheets.config.ActionPickerConfig;
import com.arthurivanets.bottomsheets.sheets.config.Config;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

/**
 * A concrete implementation of the {@link com.arthurivanets.bottomsheets.BottomSheet} which provides
 * the necessary underlying functionality to create a convenient {@link com.arthurivanets.bottomsheets.BottomSheet}-based
 * Action Picker (uses {@link ActionItem}s to represent the picker options).
 */
@SuppressLint("ViewConstructor")
public final class ActionPickerBottomSheet extends BaseActionPickerBottomSheet<ActionItem, ActionItemViewHolder, ActionPickerItemResources> {

    /**
     * Creates a new instance of the {@link ActionPickerBottomSheet} for the specified items.
     *
     * @param fragment the host fragment
     * @param items    the action items
     * @return the created {@link ActionPickerBottomSheet}
     */
    @NonNull
    public static ActionPickerBottomSheet init(@NonNull Fragment fragment,
                                               @NonNull List<ActionItem> items) {
        Preconditions.nonNull(fragment);
        Preconditions.nonNull(fragment.getActivity());
        Preconditions.nonNull(items);

        return init(fragment.getActivity(), items);
    }

    /**
     * Creates a new instance of the {@link ActionPickerBottomSheet} for the specified items.
     *
     * @param activity the host activity
     * @param items    the action items
     * @return the created {@link ActionPickerBottomSheet}
     */
    @NonNull
    public static ActionPickerBottomSheet init(@NonNull Activity activity,
                                               @NonNull List<ActionItem> items) {
        Preconditions.nonNull(activity);
        Preconditions.nonNull(items);

        return init(
            activity,
            items,
            new Config.Builder(activity).build()
        );
    }

    /**
     * Creates a new instance of the {@link ActionPickerBottomSheet} for the specified items.
     *
     * @param fragment the host fragment
     * @param items    the action items
     * @param config   the sheet configuration
     * @return the created {@link ActionPickerBottomSheet}
     */
    @NonNull
    public static ActionPickerBottomSheet init(@NonNull Fragment fragment,
                                               @NonNull List<ActionItem> items,
                                               @NonNull ActionPickerConfig config) {
        Preconditions.nonNull(fragment);
        Preconditions.nonNull(fragment.getActivity());
        Preconditions.nonNull(items);
        Preconditions.nonNull(config);

        return init(
            fragment.getActivity(),
            items,
            new ActionPickerItemResources(),
            config
        );
    }

    /**
     * Creates a new instance of the {@link ActionPickerBottomSheet} for the specified items.
     *
     * @param activity the host activity
     * @param items    the action items
     * @param config   the sheet configuration
     * @return the created {@link ActionPickerBottomSheet}
     */
    @NonNull
    public static ActionPickerBottomSheet init(@NonNull Activity activity,
                                               @NonNull List<ActionItem> items,
                                               @NonNull ActionPickerConfig config) {
        Preconditions.nonNull(activity);
        Preconditions.nonNull(items);
        Preconditions.nonNull(config);

        return init(
            activity,
            items,
            new ActionPickerItemResources(),
            config
        );
    }

    /**
     * Creates a new instance of the {@link ActionPickerBottomSheet} for the specified items.
     *
     * @param fragment      the host fragment
     * @param items         the action items
     * @param itemResources the action item resources
     * @param config        the sheet configuration
     * @return the created {@link ActionPickerBottomSheet}
     */
    @NonNull
    public static ActionPickerBottomSheet init(@NonNull Fragment fragment,
                                               @NonNull List<ActionItem> items,
                                               @NonNull ActionPickerItemResources itemResources,
                                               @NonNull ActionPickerConfig config) {
        Preconditions.nonNull(fragment);
        Preconditions.nonNull(fragment.getActivity());
        Preconditions.nonNull(items);
        Preconditions.nonNull(itemResources);
        Preconditions.nonNull(config);

        return new ActionPickerBottomSheet(
            fragment.getActivity(),
            items,
            itemResources,
            config
        );
    }

    /**
     * Creates a new instance of the {@link ActionPickerBottomSheet} for the specified items.
     *
     * @param activity      the host activity
     * @param items         the action items
     * @param itemResources the action item resources
     * @param config        the sheet configuration
     * @return the created {@link ActionPickerBottomSheet}
     */
    @NonNull
    public static ActionPickerBottomSheet init(@NonNull Activity activity,
                                               @NonNull List<ActionItem> items,
                                               @NonNull ActionPickerItemResources itemResources,
                                               @NonNull ActionPickerConfig config) {
        Preconditions.nonNull(activity);
        Preconditions.nonNull(items);
        Preconditions.nonNull(itemResources);
        Preconditions.nonNull(config);

        return new ActionPickerBottomSheet(
            activity,
            items,
            itemResources,
            config
        );
    }

    protected ActionPickerBottomSheet(@NonNull Activity context,
                                      @NonNull List<ActionItem> items,
                                      @NonNull ActionPickerItemResources itemResources,
                                      @NonNull ActionPickerConfig config) {
        super(
            context,
            items,
            itemResources,
            config
        );
    }

    @Override
    protected final int getContentViewLayoutResourceId() {
        return R.layout.bottom_sheet_base_action_picker;
    }

}

File: bottomsheets-sheets2/src/main/java/com/arthurivanets/bottomsheets/sheets2/ActionPickerBottomSheet.kt
/*
 * Copyright 2019 Arthur Ivanets, arthur.ivanets.l@gmail.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.arthurivanets.bottomsheets.sheets2

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import com.arthurivanets.bottomsheets.BaseBottomSheet
import com.arthurivanets.bottomsheets.sheets2.adapter.Item
import com.arthurivanets.bottomsheets.sheets2.adapter.SimpleRecyclerViewAdapter
import com.arthurivanets.bottomsheets.sheets2.dsl.ActionPickerConfig
import com.arthurivanets.bottomsheets.sheets2.extensions.layoutInflater
import kotlinx.android.synthetic.main.view_action_picker_bottom_sheet_content.view.*

@SuppressLint("ViewConstructor")
internal class ActionPickerBottomSheet internal constructor(
    context: Activity,
    private val pickerConfig: ActionPickerConfig
) : BaseBottomSheet(context, pickerConfig) {


    init {
        initHeader()
        initRecyclerView()
    }

    override fun onCreateSheetContentView(context: Context): View {
        return context.layoutInflater.inflate(
            R.layout.view_action_picker_bottom_sheet_content,
            this,
            false
        )
    }

    private fun initHeader() {
        pickerConfig.headerView?.let { headerView ->
            headerContainer.addView(
                headerView.getView(context, headerContainer),
                LayoutParams(
                    LayoutParams.MATCH_PARENT,
                    LayoutParams.WRAP_CONTENT
                )
            )
        }
    }

    private fun initRecyclerView() = with(recyclerView) {
        layoutManager = LinearLayoutManager(context)
        adapter = initItemAdapter()
    }

    private fun initItemAdapter(): SimpleRecyclerViewAdapter<Item<*>> {
        return SimpleRecyclerViewAdapter(
            context = context,
            items = pickerConfig.items
        ).apply {
            itemThemer = pickerConfig.itemThemeApplier
            listenerBinder = { item, holder ->
                when (item.model) {
                    is Action -> holder.itemView.setOnClickListener {
                        handleActionClick(item.model as Action)
                    }
                }
            }
        }
    }

    private fun handleActionClick(action: Action) {
        if (pickerConfig.isDismissableOnItemClick) {
            dismiss()
        }

        pickerConfig.onItemClickListener?.invoke(action)
    }

}